keps/sig-node/20191127-cri-containerstats-enhancements.md

```diff
@@ -0,0 +1,200 @@
+---
+title: Enhance ContainerStats message in CRI-API
+authors:
+  - "@fuweid"
+owning-sig: sig-node
+participating-sigs:
+  - sig-node
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-11-27
+last-updated: 2019-11-27
+status: provisional
+---
+
+# Enhance ContainerStats message in CRI-API
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+This enhancement is to extend `CpuUsage` and `MemoryUsage` fields in
+`ContainerStats`. And also add new `ProcessUsage`/`NetworkUsage` field in
+`ContainerStats`. The Container Runtime Engine will cover more metrics about
+managed container and let the metric collectors focus on node-level metrics or
+monitoring metrics.
+
+## Motivation
+
+Since more and more sandbox container runtimes shows up, like `kata-container`/
+`gVisor`/ `Firecracker`, the way to collect metrics from sandbox container is
+quite different from runC. For the third-party metric collectors, they have to
+maintain the knowledge about how to collect metrics from each kind of container
+runtimes. And more, kubelet codebase focus on core-metrics export
+(defined in [Kubernetes monitoring architecture#Terminology](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#terminology)),
+and there is no powerful standard interface for container level metrics.
+It will bring burden to end-users who wants to build powerful monitor pipeline
+(defined in [Kubernetes monitoring architecture#Executive Summary](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#executive-summary)) for sandbox container runtimes.
+
+Based on this case, we can extend Container Runtime Interface(CRI) to define
+the container level metrics. Each container runtime engine, which implemented
+CRI, knows lifecycle and metrics about each managed container better than any
+external metric collector. For the third-party metric collectors, they just
+need to follow the `ContainerStats` API defined by CRI. With standard CRI,
+it is easy for the collector from monitor pipeline (defined in [Kubernetes monitoring architecture](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#executive-summary)) to collect metrics from different kind of runtimes.
```

Comment:
Is the suggestion that metrics collectors should call the CRI stats API directly?

Reply:
Yes! Most basic metrics can be handled by Container Engine (CRI Implementation). This KEP is suggesting that metrics collectors call CRI stats directly. 

It is not good choice to use kubelet as proxy to export metrics from CRI because
1. we need to define two standard APIs for metrics, one for CRI and other one for kubelet stats API.
2. It is burden on kubelet if kubelet is as a metrics proxy. From more issues, like https://github.com/kubernetes/kubernetes/issues/68522, community is trying to reduce metrics exposed by the kubelet. 

And CRI implementation knows more details about container, for example, it knows how to get the metrics from container, no matter what kind of container is. It is easy for metrics collectors to export metrics for runC, sandbox containers and window containers.

Reply:
Thanks for the explanation.  I you should explicitly state that metrics collectors are expected to directly query the CRI.

I agree with the point about it being a burden on the kubelet to carry metrics.  I also agree that the container runtime has the best knowledge and ability to collect container metrics.

I don't agree that there will be a lot of overlap.  So far, the kubelet carries a single CPU, memory, and disk metric to support the Resource Metrics API.  This proposed change alone adds at 20 metrics...

Reply:
Sorry I didn’t make it clear. So far, kubelet only consumes few metrics you mentioned from CRI Implementation. I was saying that `IF` using kubelet to export more metrics from CRI, it will be more overlap. In order to reduce burden of kubelet, I am suggesting extending `ContainerStats` and using CRI to retrieve metrics directly. Does it make sense to you?

Reply:
I see.  It confused me that you have "Extend Core-Metrics" as a non-goal, but base your reasoning here on extending it to use the additional metrics from the CRI.  Or are you proposing adding an additional way for the kubelet to export these metrics, or using these metrics to serve the Summary API?